src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,477 @@
+        return temBAD_AMOUNT;
+
+    return preflight2 (ctx);
+}
+
+TER
+PayChanFund::doApply()
+{
+    Keylet const k (ltPAYCHAN, ctx_.tx[sfPayChannel]);
+    auto const slep = ctx_.view ().peek (k);
+    if (!slep)
+        return tecNO_ENTRY;
+
+    AccountID const src = (*slep)[sfAccount];
+    auto const txAccount = ctx_.tx[sfAccount];
+    auto const expiration = (*slep)[~sfExpiration];
+
+    {
+        auto const cancelAfter = (*slep)[~sfCancelAfter];
+        auto const closeTime =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count ();
+        if ((cancelAfter && closeTime >= *cancelAfter) ||
+            (expiration && closeTime >= *expiration))
+            return closeChannel (
+                slep, ctx_.view (), k.key, ctx_.app.journal ("View"));
+    }
+
+    if (src != txAccount)
+        // only the owner can add funds or extend
+        return tecNO_PERMISSION;
+
+    if (auto extend = ctx_.tx[~sfExpiration])
+    {
+        auto minExpiration =
+            ctx_.view ().info ().parentCloseTime.time_since_epoch ().count () +
+                (*slep)[sfSettleDelay];
+        if (expiration && *expiration < minExpiration)
+            minExpiration = *expiration;
+
+        if (*extend < minExpiration)
+            return temBAD_EXPIRATION;
+        (*slep)[~sfExpiration] = *extend;
+        ctx_.view ().update (slep);
+    }
+
+    auto const sle = ctx_.view ().peek (keylet::account (txAccount));
+
+    {
+        // Check reserve and funds availability
+        auto const balance = (*sle)[sfBalance];
+        auto const reserve =
+            ctx_.view ().fees ().accountReserve ((*sle)[sfOwnerCount]);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+
+        if (balance < reserve + ctx_.tx[sfAmount])
+            return tecUNFUNDED;
+    }
+
+    (*slep)[sfAmount] = (*slep)[sfAmount] + ctx_.tx[sfAmount];
+    ctx_.view ().update (slep);
+
+    (*sle)[sfBalance] = (*sle)[sfBalance] - ctx_.tx[sfAmount];
+    ctx_.view ().update (sle);
+
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanClaim::preflight (PreflightContext const& ctx)
+{
+    if (! ctx.rules.enabled(featurePayChan,
+            ctx.app.config().features))
+        return temDISABLED;
+
+    auto const ret = preflight1 (ctx);
+    if (!isTesSuccess (ret))
+        return ret;
+
+    auto const bal = ctx.tx[~sfBalance];
+    auto const amt = ctx.tx[~sfAmount];
+
+    if (bal && (!isXRP (*bal) || *bal <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (amt && (!isXRP (*amt) || *amt <= beast::zero))
+        return temBAD_AMOUNT;
+
+    if (bal && amt && *bal > *amt)
+        return tecNO_PERMISSION;
+
+    auto const flags = ctx.tx.getFlags ();
+    if ((flags & tfClose) && (flags & tfRenew))
+        return temMALFORMED;
+
+    if (auto const& sig = ctx.tx[~sfSignature])
```

Comment:
No reason really. I changed it to a value (<pedantic> The ref doesn't _hurt_ us here, but given the two equivalent ways I agree a value type here is expected).